libcontainer/userns/userns_linux.go

```diff
@@ -15,22 +16,33 @@ var (
 // Originally copied from github.com/lxc/lxd/shared/util.go
 func runningInUserNS() bool {
 	nsOnce.Do(func() {
-		uidmap, err := user.CurrentProcessUIDMap()
+		file, err := os.Open("/proc/self/uid_map")
 		if err != nil {
 			// This kernel-provided file only exists if user namespaces are supported
 			return
 		}
-		inUserNS = uidMapInUserNS(uidmap)
+		defer file.Close()
+
+		buf := bufio.NewReader(file)
+		l, _, err := buf.ReadLine()
+		if err != nil {
+			return
+		}
+
+		inUserNS = uidMapInUserNS(string(l))
 	})
 	return inUserNS
 }
 
-func uidMapInUserNS(uidmap []user.IDMap) bool {
+func uidMapInUserNS(uidMap string) bool {
+	var a, b, c int64
+	_, _ = fmt.Sscanf(uidMap, "%d %d %d", &a, &b, &c)
```

Comment:
Were you able to Benchmark if it makes a significant difference? It may be slightly more performance, but OTOH it's a lot more code, and the function is only executed once (in a `nsOnce.Do()`)

Reply:
Actually, I don't think `strings.SplitN` would work, because there's multiple spaces in between;

```console
/ # cat /proc/self/uid_map
         0          0 4294967295
```

So trying that, it would silently fail (on `ParseUint()`) and produce `0` for all of them; see https://play.golang.org/p/pXjr9kccesT

```go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// cat /proc/self/uuid_map
const uidMap = `         0          0 4294967295`

func main() {
	before()
	after()

}

func before() {
	var a, b, c int64

	_, _ = fmt.Sscanf(uidMap, "%d %d %d", &a, &b, &c)
	fmt.Printf("before: %d %d %d\n", a, b, c)
}

func after() {
	var a, b, c int64

	uidMapSplit := strings.SplitN(uidMap, " ", 3)
	a, _ = strconv.ParseInt(uidMapSplit[0], 10, 64)
	b, _ = strconv.ParseInt(uidMapSplit[1], 10, 64)
	c, _ = strconv.ParseInt(uidMapSplit[2], 10, 64)
	fmt.Printf("after:  %d %d %d\n", a, b, c)
}
```

Reply:
LOL, we could even do;

```go
_, err := fmt.Sscanf(strings.TrimSpace(uidMap), "0 0 4294967295")
return err == nil
```

Reply:
@thaJeztah ah dealing with redundant white-spaces will just add more code, probably `strings.Fields` instead of `strings.SplitN` should do the trick. But afaik `Fields` would not create any significant difference as compared to `Sscanf`. 
```suggestion
	uidMapSplit := strings.Fields(uidMap)
	a,_ = strconv.ParseInt(uidMapSplit[0], 10, 64)
	b,_ = strconv.ParseInt(uidMapSplit[1], 10, 64)
	c,_ = strconv.ParseInt(uidMapSplit[2], 10, 64)
```

Reply:
`strings.Fields` is 2x slower than `strings.SplitN` and that might kill all the performance improvements we gained from switching away from `fmt.Sscanf`. Without doing any benchmarks, I think we're OK using `fmt.Sscanf` here.

Reply:
Oh so I wrote some benchmarks

```go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
	"testing"
)

func benchmarkFn(b *testing.B, fn func([]byte) bool) {
	text, err := ioutil.ReadFile("/proc/self/uid_map")
	if err != nil {
		b.Fatal(err)
		return
	}
	b.ResetTimer()
	var m bool
	for i := 0; i < b.N; i++ {
		m = fn(text)
	}
	b.Log(m)
	if m != true {
		b.Fatal("wrong")
	}
}

var hostNsRegex = regexp.MustCompile(`^[[:space:]]*0[[:space:]]+0[[:space:]]+4294967295\n`)

func regexCheck(text []byte) bool {
	return hostNsRegex.Match(text)
}

func BenchmarkRegexCheck(b *testing.B) {
	benchmarkFn(b, regexCheck)
}

func sscanfCheck(text []byte) bool {
	var a, b, c int64
	n, err := fmt.Sscanf(string(text), "%d %d %d", &a, &b, &c)
	return n == 3 && err == nil && a == 0 && b == 0 && c == 4294967295
}

func BenchmarkSscanfCheck(b *testing.B) {
	benchmarkFn(b, sscanfCheck)
}

func fieldsCheck1(text []byte) bool {
	s := bytes.Fields(text)
	if len(s) != 3 {
		return false
	}
	return string(s[0]) == "0" &&
		string(s[1]) == "0" &&
		string(s[2]) == "4294967295"
}

func BenchmarkFieldsCheck1(b *testing.B) {
	benchmarkFn(b, fieldsCheck1)
}

func fieldsCheck2(text []byte) bool {
	s := strings.Fields(string(text))
	if len(s) != 3 {
		return false
	}
	return s[0] == "0" && s[1] == "0" && s[2] == "4294967295"
}

func BenchmarkFieldsCheck2(b *testing.B) {
	benchmarkFn(b, fieldsCheck2)
}

func fieldsAndParseCheck(text []byte) bool {
	split := strings.Fields(string(text))
	if len(split) != 3 {
		return false
	}
	n, err := strconv.ParseInt(split[0], 10, 64)
	if err != nil || n != 0 {
		return false
	}
	n, err = strconv.ParseInt(split[1], 10, 64)
	if err != nil || n != 0 {
		return false
	}
	n, err = strconv.ParseInt(split[2], 10, 64)
	if err != nil || n != 4294967295 {
		return false
	}

	return true
}

func BenchmarkFieldsAndParseCheck(b *testing.B) {
	benchmarkFn(b, fieldsAndParseCheck)
}
```

and the (not so obvious) results are

```console
[kir@kir-rhat ~]$ go test -bench .
goos: linux
goarch: amd64
BenchmarkRegexCheck-4            	 2160666	       539 ns/op
--- BENCH: BenchmarkRegexCheck-4
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
BenchmarkSscanfCheck-4           	  537397	      1879 ns/op
--- BENCH: BenchmarkSscanfCheck-4
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
BenchmarkFieldsCheck1-4          	 9751630	       144 ns/op
--- BENCH: BenchmarkFieldsCheck1-4
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
BenchmarkFieldsCheck2-4          	 7795519	       149 ns/op
--- BENCH: BenchmarkFieldsCheck2-4
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
BenchmarkFieldsAndParseCheck-4   	 6352525	       230 ns/op
--- BENCH: BenchmarkFieldsAndParseCheck-4
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
    a_test.go:24: true
PASS
ok  	_/home/kir	8.067s
```

So let's just keep `fmt.Sscanf`

Reply:
I also did a sscanf with strings:
```go
func sscanfStringCheck(text []byte) bool {
        var a, b, c string
        n, err := fmt.Sscanf(string(text), "%s %s %s", &a, &b, &c)
        return n == 3 && err == nil && a == "0" && b == "0" && c == "4294967295"
}

func BenchmarkSscanfStringCheck(b *testing.B) {
        benchmarkFn(b, sscanfStringCheck)
}
```

and it is slower than sscanf with ints:
```
BenchmarkSscanfCheck-4           	  747525	      2071 ns/op
BenchmarkSscanfStringCheck-4     	  817449	      1672 ns/op
```

Reply:
Now I wonder if something like `r := strings.Fields(txt)` can be replaced with `fmt.Sscanf(txt, "%s %s %s...", &r[0], &r[1], &r[2])` for cases where the number of fields are known in advance.

Reply:
Or;

```go
var a int64
_, err := fmt.Sscanf(string(bytes.TrimSpace(text)), "0 0 %d", &c)
return err == nil && a == 4294967295
```

Reply:
This version is slower :trollface: 

```go
func sscanfCheck(text []byte) bool {
        var a, b, c int64
        n, err := fmt.Sscanf(string(text), "%d %d %d", &a, &b, &c)
        return n == 3 && err == nil && a == 0 && b == 0 && c == 4294967295
}

func BenchmarkSscanfCheck(b *testing.B) {
        benchmarkFn(b, sscanfCheck)
}

func sscanf2Check(text []byte) bool {
        var a int64
        _, err := fmt.Sscanf(string(bytes.TrimSpace(text)), "0 0 %d", &a)
        return err == nil && a == 4294967295
}

func BenchmarkSscanf2Check(b *testing.B) {
        benchmarkFn(b, sscanf2Check)
}
```

Results:
```
BenchmarkSscanfCheck-4           	  744709	      2058 ns/op
BenchmarkSscanf2Check-4          	 1000000	      1237 ns/op
```

It gets better if trimspace is removed and a space is added before the first 0, but still not as fast as the original one with three %d's

Reply:
@kolyshkin Thanks a lot for these benchmarks and verifying, we cant use `SplitN` due to whitespaces so `Sscanf` should remain intact. I opened this thread but I cant see option to mark this as resolved, @thaJeztah Could you please mark this as resolved I cant see option from my side.

Reply:
It was fun to look into; and of course; https://github.com/opencontainers/runc/pull/2868#discussion_r610710079

> and the function is only executed once (in a `nsOnce.Do()`)

I'll mark it as "resolved"